core/src/main/scala/spinal/core/sim/SpinalSpec.scala

```diff
@@ -0,0 +1,273 @@
+package spinal.core.sim
+
+import org.scalatest.freespec.AnyFreeSpec
+import spinal.core.{Component, Nameable}
+
+/** Define unit / regression tests for a component */
+abstract class SpinalSpec[Dut <: Component] extends AnyFreeSpec with spinal.idslplugin.ValCallback {
+
+  override def valCallback[T](ref: T, name: String): T = {
+    ref match {
+      case nameable: Nameable => nameable.setName(name)
+      case _                  =>
+    }
+    ref
+  }
+
+  /** Override it to define the default SpinalSimConfig (else defaults to SimConfig) */
+  val config: SpinalSimConfig = SimConfig
+
+  /** `override val caching = false` to disable caching by default */
+  val caching: Boolean = true
+
+  /** `override val seed = Some(value)` to choose a fixed seed */
+  var seed: Option[Int] = None
+
+  /** Define a protocol that can be run on several benches */
+  case class Protocol[SoftDut](tests: Bench[SoftDut] => Unit) extends Nameable {
+
+    /** Run the protocol on (a) named (in a single `val`) benche(s) */
+    def runOn(benches: Bench[SoftDut]*): Unit = {
+      getName - {
+        for (bench <- benches) {
+          bench.getName - {
+            tests(bench)
+          }
+        }
+      }
+    }
+
+    /** Run the protocol on a sequence of benches */
+    def runOnAll(namedBenches: Seq[(Bench[SoftDut], String)]): Unit = {
```

Comment:
Renamed all to `runWith`/`runWithAll`

Reply:
Not sure on that: my original concern was that both functions run all the tests over the passed benches, but only one has "all" in it's name. And the difference is the names being overriden, which you couldn't tell from the names (not that you have to though, maybe just have two overrides with the same name?)

Reply:
Which is btw. also an (to me) interesting difference between `TestGroup` and `TestGroupCfg` - why does one allow for overriding the names and the other doesn´t?

Reply:
The initial reason of the `All` suffix here is that Scala considers `T*` and `Seq[T]` to finally be the same type so it prevents overloading (but `T*` cannot be given a `Seq[T]`). Maybe using a `DummyImplicit` and not having the `All` suffix would be a better solution here?

Also `runWith` is variadic (it can take one or more benches) and `runWithAll` is not, it takes a single argument which has to be a user-defined sequence, hence "`runWithAll` elements of the given sequence".

---

For `TestGroup` vs `TestGroupCfg`:

`TestGroup` can iterate on benches which are `Nameable`.

* Its `runWith` is expected to take named (put in their own `val name = ...`) benches.
* Its `runWithAll` is expected to take a sequence of tests which can be generated, so without a `val name = ...` for each bench, so it has to take the names to print additionally.

`TestGroupCfg` can iterate on configurations which are user-defined types so they can't be expected to be `Nameable`; but configurations can be converted to nice text (ex: `IdentityCfg(5)`) by default thanks to `case class`. So there is not this named vs. not named difference.

Btw the user can overload the way the text is generated from each `config`.

Reply:
I resolve this conversation as the API has changed a lot since, feel free to re-open one on the updated code if needed :smiley: